zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationPreconditions.java

```diff
@@ -330,8 +330,21 @@ private static void requireNoEventSubprocess(
       final String elementId,
       final EnumSet<BpmnEventType> allowedEventTypes,
       final String errorTemplate) {
+    final AbstractFlowElement elementById =
+        sourceProcessDefinition.getProcess().getElementById(elementId);
+
+    if (!(ExecutableActivity.class.isAssignableFrom(elementById.getClass()))) {
+      // no event subprocess event check needed because the given element cannot contain an event
+      // subprocess
+      return;
+    }
+
     final List<ExecutableStartEvent> rejectedEvents =
-        sourceProcessDefinition.getProcess().getEventSubprocesses().stream()
+        sourceProcessDefinition
+            .getProcess()
+            .getElementById(elementId, ExecutableActivity.class)
```

Comment:
@korthout do you mean `elementById` already available? It needs to be casted to `ExecutableActivity` to receive event subprocesses. I am in favour of using `getElementById` with a class definition still because it contains proper error handling. We also do the same in `requireNoBoundaryEvent()` method. WDYT?

Reply:
When it passes the `isAssignableFrom` check, you can cast it. Actually, we don't need to use `isAssignableFrom`, but we can simplify it to `instanceof` and use an instanceof pattern.

💭 I also noticed that `elementById` is just `sourceElement`
💭 I guess we can improve this in `requireNoBoundaryEvent` as well

```java
final AbstractFlowElement sourceElement =
    sourceProcessDefinition.getProcess().getElementById(elementId);

if (!(sourceElement instanceof final ExecutableActivity sourceActivity)) {
  // no event subprocess event check needed because the given element cannot contain an event
  // subprocess
  return;
}

final List<ExecutableStartEvent> rejectedEvents =
    sourceActivity.getEventSubprocesses().stream()
        .flatMap(sub -> sub.getStartEvents().stream())
        .filter(start -> !allowedEventTypes.contains(start.getEventType()))
        .toList();
```

Reply:
I like the notation above, it is a lot simpler 👍